Main Menu Tiles - Fall Documentation 2019 (#141)
smartdevicelink/sdl_app_library_guides: docs/Displaying a User Interface/Main Menu/index.md

Every template has a main menu button. The position of this button varies between templates and cannot be removed from the template. Some OEMs may format certain templates to not display the main menu button if you have no menu items (such as the navigation map view).
!!!

> ## Setting a Menu Style

Reviewer: Move the **Menu Template** header and screenshot that is currently above this section to the **Setting the Menu Layout** section. Also add a generic hmi screenshot of the same menu with the tile format.

> ## Setting the Menu Layout

Reviewer 2: ```suggestion
## Setting the Menu Layout (RPC v6.0+)
```

> Core v6.0+ supports displaying menu items as tiles or as a list depending on OEM. To see if the display supports menu tiles check @![iOS] `self.sdlManager.systemCapabilityManager.defaultMainWindowCapability'`s `menuLayoutsAvailable` !@@![android,javaEE,javaSE] `//ToDo - add info`!@ property. To set a menu layout please set the `screenManager`s  `menuConfiguration` property. The default layout is `list` if none is set.

Reviewer: ```suggestion
If connected to a newer head unit (RPC v.6.0+) you may have the option to display menu items as a grid of tiles in addition to the default list layout. To find out if the display supports menu tiles, check the system capability manager's @![iOS]`defaultMainWindowCapability.menuLayoutsAvailable`!@@![android,javaEE,javaSE] `getDefaultMainWindowCapability().getMenuLayoutsAvailable()`!@ property after successfully connecting to the head unit. To set the menu layout using the screen manager, you will need to set the `screenManager`s `menuConfiguration` property.
```

> On some newer head units, you may have the option to display menu items as a grid of tiles instead of the default list layout. To determine if the head unit supports the tiles layout, check the system capability manager's @![iOS]`defaultMainWindowCapability.menuLayoutsAvailable`!@@![android,javaEE,javaSE] `getDefaultMainWindowCapability().getMenuLayoutsAvailable()`!@ property after successfully connecting to the head unit. To set the menu layout using the screen manager, you will need to set the `screenManager`s `menuConfiguration` property.

Reviewer 2: ```suggestion
On some newer head units, you may have the option to display menu items as a grid of tiles instead of the default list layout. To determine if the head unit supports the tiles layout, check the `SystemCapabilityManager`'s @![iOS]`defaultMainWindowCapability.menuLayoutsAvailable`!@@![android,javaEE,javaSE] `getDefaultMainWindowCapability().getMenuLayoutsAvailable()`!@ property after successfully connecting to the head unit. To set the menu layout using the screen manager, you will need to set the `ScreenManager.menuConfiguration` property.
```

> If connected to a newer head unit (RPC v.6.0+) you may have the option to display menu items as a grid of tiles in addition to the default list layout. To find out if the display supports menu tiles, check the system capability manager's @![iOS]`defaultMainWindowCapability.menuLayoutsAvailable`!@@![android,javaEE,javaSE] `getDefaultMainWindowCapability().getMenuLayoutsAvailable()`!@ property after successfully connecting to the head unit. To set the menu layout using the screen manager, you will need to set the `screenManager`s `menuConfiguration` property.

Reviewer 2: ```suggestion
On some newer head units, you may have the option to display menu items as a grid of tiles instead of the default list layout. To determine if the head unit supports the tiles layout, check the system capability manager's @![iOS]`defaultMainWindowCapability.menuLayoutsAvailable`!@@![android,javaEE,javaSE] `getDefaultMainWindowCapability().getMenuLayoutsAvailable()`!@ property after successfully connecting to the head unit. To set the menu layout using the screen manager, you will need to set the `screenManager`s `menuConfiguration` property.
```

!!! Note

> If `menuConfiguration` is set after a menu already exists, sub-menu layouts will not be updated.  A new menu will have to be set to see the new sub-menu layout.

Author: from the inline documentation in `SDLScreenManger` line 127

> If `menuConfiguration` is set after a menu already exists, sub-menu layouts will not be updated.  A new menu will have to be set to see the new sub-menu layout.

Author: also right here. https://github.com/smartdevicelink/generic_hmi/blob/develop/src/js/reducers.js#L319

!!!

@![iOS]
##### Objective-C
```objc

> SDLMenuConfiguration *menuConfiguration = [SDLMenuConfiguration alloc] initWithMainMenuLayout:<#MenuLayout#> defaultSubmenuLayout:<#MenuLayout#>];

Reviewer: ```suggestion
SDLMenuConfiguration *menuConfiguration = [[SDLMenuConfiguration alloc] initWithMainMenuLayout:<#SDLMenuLayout#> defaultSubmenuLayout:<#SDLMenuLayout#>];
```

> self.sdlManager.screenManager.menuConfiguration = menuConfiguration 

Reviewer: ```suggestion
self.sdlManager.screenManager.menuConfiguration = menuConfiguration;
```

```
##### Swift
```swift

> let menuConfiguration = SDLMenuConfiguration(mainMenuLayout: <#MenuLayout#>, defaultSubmenuLayout: <#MenuLayout#>)

Reviewer: ```suggestion
let menuConfiguration = SDLMenuConfiguration(mainMenuLayout: <#SDLMenuLayout#>, defaultSubmenuLayout: <#SDLMenuLayout#>)
```

!@

### Adding Submenus

> Adding a submenu is as simple as adding subcells to a @![iOS]`SDLMenuCell`!@ @![android, javaSE, javaEE]`SdlMenuCell`!@. The submenu is automatically displayed when selected by the user. Currently menus only support one layer of subcells. In Core v6.0+ it is possible to set individual submenus to use different layouts such as tiles or lists.

Reviewer: ```suggestion
Adding a submenu is as simple as adding subcells to a @![iOS]`SDLMenuCell`!@ @![android, javaSE, javaEE]`SdlMenuCell`!@. The submenu is automatically displayed when selected by the user. Currently menus only support one layer of subcells. In RPC v6.0+ it is possible to set individual submenus to use different layouts such as tiles or lists.
```

@![iOS]
##### Objective-C

> ###### Pre Core v6.0

Reviewer: We can just update the code snippets to use the new `init`s as it will still work with head units with pre RPC v6.0. The the tile layout will just not be set.

!@

### Adding Submenus

> Adding a submenu is as simple as adding subcells to a @![iOS]`SDLMenuCell`!@ @![android, javaSE, javaEE]`SdlMenuCell`!@. The submenu is automatically displayed when selected by the user. Currently menus only support one layer of subcells. In RPC v6.0+ it is possible to set individual submenus to use different layouts such as tiles or lists.

Reviewer 2: ```suggestion
Adding a submenu is as simple as adding subcells to a @![iOS]`SDLMenuCell`!@@![android, javaSE, javaEE]`SdlMenuCell`!@. The submenu is automatically displayed when selected by the user. Currently menus only support one layer of subcells. In RPC v6.0+ it is possible to set individual submenus to use different layouts such as tiles or lists.
```

SDLMenuCell *submenuCell = [[SDLMenuCell alloc] initWithTitle:<#NSString#> icon:<#SDLArtwork?#> subCells:@[cell]];
self.sdlManager.screenManager.menu = @[submenuCell];
```

> ###### Core v6.0+

Reviewer: ```suggestion
###### RPC v6.0+
```

```objc
// Create the inner menu cell
SDLMenuCell *cell = [[SDLMenuCell alloc] initWithTitle:<#NSString#> icon:<#SDLArtwork#> voiceCommands:<#@[NSString]#> handler:^(SDLTriggerSource  _Nonnull triggerSource) {
    // Menu item was selected, check the `triggerSource` to know if the user used touch or voice to activate it
    <#Handle the cell's selection#>
}];

// Create and set the submenu cell

> SDLMenuCell *submenuCell = [[SDLMenuCell alloc] initWithTitle:<#NSString#> icon:<#SDLArtwork?#> submenuLayout:<#Menu Layout#>, subCells:@[cell]];

Reviewer: ```suggestion
SDLMenuCell *submenuCell = [[SDLMenuCell alloc] initWithTitle:<#NSString#> icon:<#SDLArtwork?#> submenuLayout:<#SDLMenuLayout#>, subCells:@[cell]];
```

self.sdlManager.screenManager.menu = @[submenuCell];
```

##### Swift

> ###### Pre Core v6.0

Reviewer: We can just update the code snippets to use the new `init`s as it will still work with head units with pre RPC v6.0. The the tile layout will just not be set.

```swift
// Create the inner menu cell
let cell = SDLMenuCell(title: <#String#>, icon: <#SDLArtwork?#>, voiceCommands: <#[String]?#>) { (triggerSource: SDLTriggerSource) in
    // Menu item was selected, check the `triggerSource` to know if the user used touch or voice to activate it
    <#Handle the cell's selection#>
}

let submenuCell = SDLMenuCell(title: <#String#>, icon: <#SDLArtwork?#>, subCells: [cell])
sdlManager.screenManager.menu = [submenuCell]
```
###### Core v6.0+
```swift
// Create the inner menu cell
let cell = SDLMenuCell(title: <#String#>, icon: <#SDLArtwork?#>, voiceCommands: <#[String]?#>) { (triggerSource: SDLTriggerSource) in
    // Menu item was selected, check the `triggerSource` to know if the user used touch or voice to activate it
    <#Handle the cell's selection#>
}

> let submenuCell = SDLMenuCell(title: <#String#>, icon: <#SDLArtwork?#>, submenuLayout: <#Menu Layout#>, subCells: [cell])

Reviewer: ```suggestion
let submenuCell = SDLMenuCell(title: <#String#>, icon: <#SDLArtwork?#>, submenuLayout: <#SDLMenuLayout#>, subCells: [cell])
```

sdlManager.getScreenManager().setMenu(Collections.singletonList(cell));
```
```//ToDo add new example to set a menu layout.
```
!@

> ### Artwork

Reviewer 2: ```suggestion
### Menu Item Artwork
```

sdlManager.getScreenManager().setMenu(Collections.singletonList(cell));
```

> ```//ToDo add new example to set a menu layout.

Author: fixed
